runtime/test/multiplexer-test.js

```diff
@@ -67,4 +70,134 @@ describe('Multiplexer', function() {
+          consume item
+      `
+    });
+    const renderParticle = manifest.particles[0];
+    const renderParticleSpec = JSON.stringify(renderParticle.toLiteral());
+    const renderRecipe = DomParticle
+                             .buildManifest`
+${renderParticle}
+recipe
+  use '{{item_id}}' as v1
+  slot '{{slot_id}}' as s1
+  ${renderParticle.name}
+    ${renderParticle.connections[0].name} <- v1
+    consume item as s1
+  `.trim();
+    return {renderParticleSpec, renderRecipe: renderRecipe.replace(/[\n]/g, '\\n')};
+  };
+
+  it('varies render particle across heterogeneous items with embedded recipes', async () => {
+    let resultA = await buildEmbeddedItemValues('A');
+    let resultB = await buildEmbeddedItemValues('B');
+
+    const manifest = await Manifest.parse(`
+import 'shell/artifacts/Common/List.manifest'
+
+schema Frob
+  Text renderParticleSpec
+  Text renderRecipe
+
+store Frobs of [Frob] in FrobsJson
+ resource FrobsJson
+   start
+   [
+     {"renderParticleSpec": ${JSON.stringify(resultA.renderParticleSpec)},
+      "renderRecipe": "${resultA.renderRecipe}"},
+     {"renderParticleSpec": ${JSON.stringify(resultB.renderParticleSpec)},
+      "renderRecipe": "${resultB.renderRecipe}"}
+    ]
+
+particle FrobMuxer in 'runtime/test/artifacts/transformations/frob-muxer.js'
+  FrobMuxer(in [Frob] list)
+  consume set of item
+
+recipe
+  use Frobs as frobs
+  List
+    items = frobs
+  FrobMuxer
+    list = frobs
+    `, {loader, fileName: './manifest.manifest'});
+
+    let recipe = manifest.recipes[0];
+    let slotComposer = new MockSlotComposer();
+    slotComposer
+      .newExpectations()
+      .expectRenderSlot('List', 'root', {contentTypes: ['template', 'model']})
+      .expectRenderSlot('A', 'item', {contentTypes: ['template', 'model']})
+      .expectRenderSlot('A', 'item', {contentTypes: ['template', 'model']})
+      .expectRenderSlot('FrobMuxer', 'item', {contentTypes: ['template', 'model']})
+      .expectRenderSlot('A', 'item', {contentTypes: ['template', 'model']})
+      // TODO(wkorman): We'd like to assert that the template for the last two
+      // calls to FrobMuxer contains both templates from A and B particles, and
+      // similarly that the template for A and B above is the template for each
+      // of the respective particles, but it appears we have to enhance
+      // MockSlotComposer support for repeated expectations for the same
+      // particle with verifications first. If we uncomment the `verify`
+      // bits below we get failures that seem to show the verification is
+      // applied even to previous expectations.
+      .expectRenderSlot('FrobMuxer', 'item', {contentTypes: ['template', 'model']
+        })
+      .expectRenderSlot('B', 'item', {contentTypes: ['template', 'model']})
+      .expectRenderSlot('B', 'item', {contentTypes: ['template', 'model']})
+      .expectRenderSlot('FrobMuxer', 'item', {contentTypes: ['template', 'model']
+      // ,
```

Comment:
is this section commented out temporary? or should it be verifiable, but not still working? maybe add a todo?

Reply:
It's temporary, there's a comment above, I added one here and below referencing that one. I think there's a bug (or missing/revised feature behavior) in MockSlotComposer around verifying. If we only want to verify say the last of a sequence of expectations, the verify hook for that last one is called for all of the preceding ones as well. So, I've left something at least close to what we likely want commented out and will file an issue to track.

Reply:
The `verify` method shouldn't assert, it should only return true or false, depending on whether it is happy. This way it ignores all preceding calls, until the expectation is finally satisfied and resolved (if `verify` never returned true, the expectations remains unresolved and the test fails).

Reply:
Ah, I see. OK, so in this test we see four calls to FrobMuxer. I can arrange for a rough verification that passes for the third call, with:

```
    const verifyTemplateAB = (content) => {
      if (Object.keys(content.template).length == 2 &&
          Object.values(content.template)[0] == '<div mux-a><{{value}}</div>' &&
          Object.values(content.template)[1] == '<div mux-b><{{value}}</div>') {
        return true;
      }
    };
```

but then we fail on the fourth FrobMuxer render call, with:

`(node:29464) UnhandledPromiseRejectionWarning: AssertionError: Got a renderSlot from FrobMuxer:item (content types: template, model), but not expecting anything further.`

If I try to add the verify to the fourth expectation as well, then we get:

`AssertionError: Unspecified AssertionError at MockSlotComposer._addRenderExpectation (file://runtime/test/mock-slot-composer.js:180:7)`

because we already have the completed verify from the previous call. If I only put the verify on the third expectation, we still fail on the fourth expectation with the `not expecting anything further.` error above.

Maybe I'm trying to use MockSlotComposer wrong? I thought maybe it wasn't too crazy to work on adding the verify bits in a separate PR given the challenges but you may be able to steer me more quickly.

Reply:
I suppose we could have the verify function only return true once it sees those 2x...

Reply:
Is the 4th call different from the 3rd one though? 
Some possibly helpful debugging tips: 
- MockSlotComposer has `detailedLogDebug` that might be helpful in figuring out what are the remaining non-satisfied expectations at any given time. 
- You can uncommented a commented out console.log in `MockSlotComposer::renderSlot` method to have all the render calls listed.

It is very much possible that MockSlotComposer needs some additional verification support.  Let me know, if you'd rather i patch the PR and debug myself.

Reply:
I think the 3rd and 4th are the same. I just had the three verify hooks all expect to see the matching template content the appropriate number of times for now. I only really care about validating it once, so you can see that if MockSlotComposer behaved differently (such as only calling the verify hook for the specific expectation or expectation-index with which it was registered) it would simplify the verify hook logic itself.

Also obviously there may be some optimization opportunities with templates/muxing since we see these repeated calls. I have https://github.com/PolymerLabs/arcs/issues/1233 open to look further into the template optimization subsequently since this PR essentially disables it for now.

Reply:
alternatively you could pass {strict: false} to MockSlotComposer ctor - it'll log warning instead of asserting for unexpected render calls. i actually think in this case it is better, because you get to actually verify that both templates are passed, which is basically the purpose of this..

Reply:
OK, reworked to use non-strict, seems simpler, thanks.